mesonbuild/interpreter.py

```diff
@@ -2754,6 +2754,15 @@ def do_subproject(self, dirname: str, method: str, kwargs):
 
         default_options = mesonlib.stringlistify(kwargs.get('default_options', []))
         default_options = coredata.create_options_dict(default_options)
+
+        # We want to overwrite the keyword arguments with options from the
+        # cross/native file. Take the parent options first, then override that
+        # with the subproject specific options
+        for n in ['', dirname]:
+            default_options.update(self.environment.meson_options.host.get(n, {}))
+            default_options.update({'build.{}'.format(k): v for k, v in self.environment.meson_options.build.get(n, {}).items()
+                                    if k in coredata.builtin_options_per_machine})
```

Comment:
You prepend `build.` here and also in set_default_options(). Here you set them in default_options, but then you also set them in set_default_options()... I'm lost.

The code in set_default_options() and do_subproject() is already waaay to complicated, I would **STRONGLY** recommend refactoring your code to update self.cmd_line_options in environment.py as I did in myhttps://github.com/mesonbuild/meson/pull/7293. That simplify a lot the implementation of this PR and makes it obvious that options in machine files are handled  the same way as if passed on the command line.

Reply:
I think when @Ericson2314 and I talked about this (it was a while ago on IRC) we were talking about removing env.cmd_line_options all together and using the individual options instead.

Reply:
I don't know what's wrong with `env.cmd_line_options` and I would like to discuss it if there are issues raised since I'm the one who wrote that code ;-) But in any case, that's not a reason to make the code more complicated than it could be in the meantime.

btw, I'm not saying your code is wrong. It does the right thing in all the tests I've done, I just think it could do the same thing in a simpler and more maintainable way. From past experience I know that set_default_options() is a can of worms with many subtle edge cases, better avoid messing with it unless it's for rewriting that mess :D